CFC_DataCollector/modeinfer/pipeline.py

```diff
@@ -163,51 +202,69 @@ def getGroundTruthMode(self, section):
 # 19. both start and end close to bus stop
 # 20. both start and end close to train station
 # 21. both start and end close to airport
+#TODO: pass appropriate vector (featureMatrix[i]) rather than Matrix and index..
   def updateFeatureMatrixRowWithSection(self, featureMatrix, i, section):
-    featureMatrix[i, 0] = section['distance']
-    featureMatrix[i, 1] = (section['section_end_datetime'] - section['section_start_datetime']).total_seconds()
-
-    # Deal with unknown modes like "airplane"
-    try:
-      featureMatrix[i, 2] = section['mode']
-    except ValueError:
-      featureMatrix[i, 2] = 0
-
-    featureMatrix[i, 3] = section['section_id']
-    featureMatrix[i, 4] = calAvgSpeed(section)
-    speeds = calSpeeds(section)
-    if speeds != None:
-        featureMatrix[i, 5] = np.mean(speeds)
-        featureMatrix[i, 6] = np.std(speeds)
-        featureMatrix[i, 7] = np.max(speeds)
-    else:
-        # They will remain zero
-        pass
-    accels = calAccels(section)
-    if accels != None and len(accels) > 0:
-        featureMatrix[i, 8] = np.max(accels)
-    else:
-        # They will remain zero
-        pass
-    featureMatrix[i, 9] = ('commute' in section) and (section['commute'] == 'to' or section['commute'] == 'from')
-    featureMatrix[i, 10] = calHCR(section)
-    featureMatrix[i, 11] = calSR(section)
-    featureMatrix[i, 12] = calVCR(section)
-    if 'section_start_point' in section and section['section_start_point'] != None:
-        startCoords = section['section_start_point']['coordinates']
-        featureMatrix[i, 13] = startCoords[0]
-        featureMatrix[i, 14] = startCoords[1]
-    
-    if 'section_end_point' in section and section['section_end_point'] != None:
-        endCoords = section['section_end_point']['coordinates']
-        featureMatrix[i, 15] = endCoords[0]
-        featureMatrix[i, 16] = endCoords[1]
-    
-    featureMatrix[i, 17] = section['section_start_datetime'].time().hour
-    featureMatrix[i, 18] = section['section_end_datetime'].time().hour
-    
-    featureMatrix[i, 19] = mode_start_end_coverage(section, self.bus_cluster,105)
-    featureMatrix[i, 20] = mode_start_end_coverage(section, self.train_cluster,600)
+    points = section["track_points"]
+    if len(points) > 4:
+        points = np.array([[(datetime.datetime.strptime(point['time'].split('-')[0], "%Y%m%dT%H%M%S") - datetime.datetime(1970,1,1)).total_seconds(), 
+            point["track_location"]["coordinates"][0], point["track_location"]["coordinates"][1]] for point in points])
+        time_stamp = points[:,0].reshape(len(points[:,0]), 1)
+        lon_lat = points[:,1:]
+        model_ransac = linear_model.RANSACRegressor(linear_model.LinearRegression())
```

Comment:
Can you provide some intuition around why we would want to put the RANSAC regression code into the machine learning pipeline?

This seems to have two problems:

1.  It seems to me that you would be running RANSAC over and over on the same set of trips. For example, consider the scenario in which I took 3 trips (t1, t2, t3) today and classified them today.
  - Then, (t1, t2 and t3) would become part of the training set tomorrow. So I would run RANSAC on them then.
  - Then, they would continue to be part of the training set day after tomorrow, so I would run RANSAC on them again.
2. Running RANSAC at this point will leave the values in the database untouched. This will mean that the users will see the zigzag points as well. Having zigzag, unsmoothed points in the map displayed to the user makes it harder for users to classify trips, and to have faith in our data collection.

Running RANSAC at the point that the data comes in from the phone (`tripManager.storeSensedTrips`) seems like it would avoid both of these problems, which is why I suggested it in the issue description.